pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -56,9 +56,27 @@ type MatrixInclude struct {
 // Combination is a map, mainly defined to hold a single combination from a Matrix with key as param.Name and value as param.Value
 type Combination map[string]string
 
-// Combinations is a list of combinations
+// Combinations is a Combination list
 type Combinations []Combination
 
+// FanOut generates Combinations, which is a list of mapped Combination by fanning out matrix
+// Parameters including Matrix Include Parameters
+func (m *Matrix) FanOut() Combinations {
+	var combinations Combinations
+	// If only Matrix Include Parameters exists, generate and return explicit Combinations
+	if m.hasInclude() && !m.hasParams() {
+		return m.fanOutExplictCombinations()
+	}
+	// Fan out Matrix Parameters
+	for _, parameter := range m.Params {
+		combinations = combinations.fanOut(parameter)
+	}
```

Comment:
until this point, we have collected `combinations` based on `matrix.params` but in case of a `matrix` with `params` and `include` both sections specified, the `combinations` generated from `include` section are not persisted i.e. combinations from `m.fanOutExplictCombinations()`

If we persist combinations generated from `include` section, we can compare `combinations` from `include` section with the `combinations` from `params` section and replace the `values` in place.

Reply:
Thanks Priti! I've incorporated these changes.